### ISSUE TITLE
react-redux with mapState and pure

### ISSUE THREAD

**Issue by maxsalven** [flowtype/flow-typed#1946]
A huge thank you to @villesau for all your work on this!

I think I found a small issue:
```
// @flow

import { connect } from "react-redux";

type Props = {
  stringProp: string,
};

const Component = ({ stringProp }: Props) => {
  return stringProp;
};

const mapStateToProps = state => ({
  stringProp: false,
});
```

This correctly tells me that my stringProp is not a string:
```
export default connect(mapStateToProps)(Component);
```

However, this does not give any Flow errors:
```
export default connect(mapStateToProps, null, null, { pure: false })(Component);
```

I manually added a very quick workaround for this specific situation, but would be great to get 'proper' support for this in the libdef.

```
  declare export function connect<Com: ComponentType<*>, DP: Object, RSP: Object>(
    mapStateToProps: MapStateToProps<DP, RSP>,
    mapDispatchToProps?: null,
    mergeProps?: null,
    options: any
  ): (component: Com) => ComponentType<$Diff<OmitDispatch<ElementConfig<Com>>, RSP> & DP>;
```

**Comment by maxsalven** [comment]
Strangely the first example `export default connect(mapStateToProps)(Component);` is no longer giving me an error(?). Unfortunately I'm out of time to investigate, before I dive further @villesau could you please let me know if it correctly gives you an error, thanks.

**Comment by villesau** [comment]
@maxsalven It's not giving you any error if you are not using the component anywhere :) In that case flow perhaps _could_ know that type of `stringProp`s contradicts even without using the component anywhere but apparently Flow doesn't type check it if some of the property 'sources' is missing (component usage in this case).

**Comment by maxsalven** [comment]
Not sure I follow, the component needs to be used externally to be properly type checked? This seems opposite to how Flow currently works, I can define a React component that isn't used anywhere, and I'll still get Flow errors.

When I'm writing a new component, seeing Flow errors whilst writing the component is extremely important? Maybe I'm alone here, but if it's the case that the `connect` call doesn't get type checked until I've actually got the component being used somewhere else, then that's a huge regression for my use cases unfortunately.

As reference, the old version of the tool, with all its shortcomings, correctly reports an error for this file which is completely standalone:

```
// @flow

import { connect } from "react-redux";
import type { Connector } from "react-redux";

type ContainerProps = {};

type Props = ContainerProps & {
  stringProp: string,
};

const Component = ({ stringProp }: Props) => {
  return stringProp;
};

const mapStateToProps = (state): * => ({
  stringProp: 5,
});

const connector: Connector<ContainerProps, Props> = connect(mapStateToProps);

const ConnectedComponent = connector(Component);
```

```
Cannot assign connect(...) to connector because number [1] is incompatible with string [2] in property stringProp of
type argument P [3].
```

Do you think there's a way to achieve this behavior with the new types or is it simple not possible given the way they are structured? Thanks!

**Comment by AndrewSouthpaw** [comment]
So, I was suuuuper confused about it as well, that you have to use it before it type checks. It's a bit unusual, and also I think it doesn't actually matter. If you're not actually using the component, then whether it type-checks or not is... not as critical. And, though I don't know the details of what @villesau did, it seems like type checking on usage is essential to the success of the new definition, and intelligently determining props based on what's passed in, etc.

**Comment by villesau** [comment]
To be honest, not using component didn't even crossed in my mind :D In normal use cases component is always used _somewhere_ If you are writing ui-library, then you at least have tests using it.

Luckily I found a potential solution for that. The current solution is built on top of returned component prop types. But there are also couple of other places to put the 'logic', like type parameters. I need to experiment a bit to figure out wether there are any downsides or not.

@maxsalven out of curiosity, can you describe your use case where you build components but don't use them?

**Comment by maxsalven** [comment]
Thanks all! The use case is whenever I am writing a new component, it will be odd to have that component type check successfully whilst I am actually doing the coding, and then only after I utilize it elsewhere, will I realize I made a bunch of mistakes whilst writing it.

It's not the end of the world by any means, especially now that I know what is going on! If your POC works then fantastic, but if not perhaps some sort of comment / docs in the libdef would be helpful.

Thanks again, I truly do appreciate the work that has gone into this.

**Comment by villesau** [comment]
@maxsalven I have a PR open for the non-used-component-problem. Now when looking at your problem with options I realized that current libdefs assumes you to have mergeProps in place in order to be able to use options. Your workaround is very close to what it is supposed to be: All the other definitions should have something like `connect(...mapXtoProps, mergeProps?: null, options?: ?ConnectOptions<A, B, C, D>)`

**Comment by adamterlson** [comment]
> If you're not actually using the component, then whether it type-checks or not is... not as critical.

This is indeed a rather large issue for my use case, where `connect` is one of many higher order components that ultimately receive a component.

Consider what used to work for us:

```javascript
type OwnProps = { ... }
type PropTypes = { ... }

const MyComponent = (props: PropTypes) => ....

const connector: Connector<OwnProps, PropTypes> = connect(...)

export default compose(
    connector,
    hoc1,
    hoc2,
    hoc3,
)(MyComponent)
```

**Comment by maxsalven** [comment]
Thanks @villesau, the new libs are working great. However I think the original issue (`pure: false` without mergeProps – `export default connect(mapStateToProps, null, null, { pure: false })(Component);`) is still outstanding?

**Comment by villesau** [comment]
@maxsalven yes. Would you be able to do a PR? I think your suggested solution is pretty good.

**Comment by pajasevi** [comment]
I have a similar problem.
```
export default connect(mapStateToProps, null, null, { pure: false })(UserInfo);
```
Then I use this component somewhere else:
```
<UserInfo asButton />
```

And it gives me this error: `Error:(74, 47) Cannot create 'UserInfo' element because props [1] is incompatible with empty [2].`

Both are standard React Components and cannot be pure because of React Router's `Navlink` component.
